### PR TITLE
fix: initialData & placeholderData does not enforce output types

### LINKED ISSUE #4486: bug: wrong types for initialData
### Provide environment information

```
  System:
    OS: Linux 5.15 Debian GNU/Linux 11 (bullseye) 11 (bullseye)
    CPU: (2) x64 AMD EPYC
    Memory: 1.08 GB / 2.01 GB
    Container: Yes
    Shell: 5.1.4 - /bin/bash
  Binaries:
    Node: 16.17.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 8.15.0 - /usr/local/bin/npm
    Watchman: 20220528.183901.0 - /usr/local/bin/watchman
  npmPackages:
    @tanstack/react-query: 4.29.12 => 4.14.5 
    @trpc/client: 10.29.1 => 10.6.0 
    @trpc/next: 10.29.1 => 10.6.0 
    @trpc/react-query: 10.29.1 => 10.6.0 
    @trpc/server: 10.29.1 => 10.6.0 
    next: ^12.3.1 => 12.3.2 
    react: ^18.2.0 => 18.2.0 
    typescript: ^4.8.3 => 4.8.4 
```

### Describe the bug

it seems that I can pass in anything into the `initialData` field and typescript doesn't complain.

### Link to reproduction

https://codesandbox.io/p/sandbox/immutable-grass-wuw9r2?file=%2Fsrc%2Fpages%2Findex.tsx%3A13%2C26

### To reproduce

in the provided sandbox, I'm passing `initialData: 5` to the standard greeting router, which actually returns `{ text: string }`. I'd expect there to be a type error when passing the wrong initialData.

Note that this works fine with raw react-query, as you can see in this [typescript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA2iSxwDecArqlgIptZQCecAL5wiJcgAEYyAHappuANYB6SnhgBaAI68BtGnDi4Ic+AHNKWGMBlm4AXnaceffgApmBw3B2uA0lj8AFxwANpkMFjyZAC6ADRehr4CAGIyIW4AlA4AfHAACsQgwJwAdJSoEAA2AG5YHnCRAB4wIWQAFlhVVRBwAO7QVQAmZMKZCd5wyspwgLwbgBV7je0lcCt9xLaJqzLA1shVACLI0iEArBPe8vtYACqgWCEAkjIENrv8F0KZANxAA). The error is weird because of our overloads, which will get better in v5 (playground with alpha would be [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA2iSxwDecArqlgIptZQCecAL5wiJcgAEYyAHappuANYB6SnhgBaAI68BZOMuVwY-MFlQAuFABswAC2R04cXBDnwA5pSwxgMj3AAvOycPHz8ABTMNM7OOuEA0lj8VgDaZDDmMGQAugA0MbHxAgBiMlYRAJRBAHxwAArEIMCcAHSUqBDWAG5YUcZYAB4wVmR2WNbWEHAA7tDWACb6QpUFsQZGgLwbgBV7xnYtcAczxP6Fzn7AvsjWACLI0lYArGux8tdYACqgWFYAkjIEC4mF4rADcQA)) 😅 

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

<sub>[T-104](https://linear.app/trpc/issue/T-104/bug-wrong-types-for-initialdata)</sub>

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes #4486

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

## ✅ Checklist

- [ ] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Review by KATT — APPROVED** [review]
🐐

File: packages/react-query/src/createTRPCReact.tsx [hunk 1]
```diff
@@ -53,8 +53,8 @@ export interface ProcedureUseQuery<
   TPath extends string,
 > {
   <
-    TQueryFnData = inferTransformedProcedureOutput<TProcedure>,
-    TData = inferTransformedProcedureOutput<TProcedure>,
+    TQueryFnData extends inferTransformedProcedureOutput<TProcedure> = inferTransformedProcedureOutput<TProcedure>,
+    TData extends inferTransformedProcedureOutput<TProcedure> = inferTransformedProcedureOutput<TProcedure>,
```
Comment (bdsqqq) [inline packages/react-query/src/createTRPCReact.tsx:57]: @juliusmarminge  What's the difference between just assigning this compared to extending + assigning? 🤔

**Review by bdsqqq — COMMENTED** [review]
🐐

File: packages/react-query/src/createTRPCReact.tsx (hunk 1 above)
Comment (juliusmarminge) [reply packages/react-query/src/createTRPCReact.tsx:57]: @Nick-Lucas calls it the `extends =` pattern but it essentielly lets you have it as a generic (so others can override it) whilst keeping the constraint

**Comment by TkDodo** [comment]
thanks so much @juliusmarminge - I can confirm that both things reported are fixed by this 🙌

**Pull request merged**